storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,418 @@
+  /**
+   * @brief Extracts a comparison predicate into an SMAPredicate.
+   *
+   * @param predicate A comparison of the form {attribute} {comparison} {literal}
+   *                  or {literal} {comparison} {attribute}.
+   * @return An SMAPredicate pointer which the caller must manage.
+   */
+  static SMAPredicate* ExtractSMAPredicate(const ComparisonPredicate& predicate);
+
+  const attribute_id attribute;
+  const ComparisonID comparison;
+  const TypedValue literal;
+  // How much of the TupleStore this predicate will select. kUnsolved indicates
+  // that the predicate had not been used previously.
+  Selectivity selectivity;
+
+ private:
+  SMAPredicate(const attribute_id attr,
+               const ComparisonID comp,
+               const TypedValue lit)
+      : attribute(attr),
+        comparison(comp),
+        literal(lit),
+        selectivity(Selectivity::kUnsolved) { }
+};
+
+//// Components of the index.
+
+// A 64-bit header.
+struct SMAHeader {
+  // Count refers the SQL aggregate COUNT.
+  std::uint32_t count;
+
+  union {
+    // Consistent refers to the index being in a consistent state.
+    bool consistent;
+    // Padding is so we keep this data structure to 64 bits in length.
+    std::uint32_t padding;
+  };
+};
+
+// Reference to an attribute value in a tuple. This struct is used to keep a
+// of the min and max value for an attribute.
+struct EntryReference {
+  tuple_id tuple;
+  // There are certain cases when the entry reference can be invalid such as
+  // when the index is being rebuilt. value should not be used in this case.
+  bool valid;
+  // A copy of the typed value referenced by tuple.
+  TypedValue value;
+};
+
+// Index entry for an attribute in the SMA.
+struct SMAEntry {
+  attribute_id attribute;
+  // The type of the attribute contained in this entry.
+  TypeID type;
+  // A reference to the minimum value of the attribute.
+  EntryReference min_entry;
+  // A reference to the maximum value of the attribute.
+  EntryReference max_entry;
+  // The sum of the attribute. This is only used in the case of numeric types.
+  TypedValue sum;
+};
+
+}  // namespace sma_internal
+
+
+/**
+ * @brief Small Materialized Aggregate SubBlock.
+ * @details Keeps account of several types of aggregate functions per Block.
+ *          Currently supports min, max, sum, and count.
+ *          Physical organization of the block is as follows:
+ *
+ *          [SMAHeader][SMAEntry x number of indexed attributes]
+ */
+class SMAIndexSubBlock : public IndexSubBlock {
+ public:
+  SMAIndexSubBlock(const TupleStorageSubBlock &tuple_store,
+                   const IndexSubBlockDescription &description,
+                   const bool new_block,
+                   void *sub_block_memory,
+                   const std::size_t sub_block_memory_size);
+
+  /**
+   * @brief Frees data associated with variable length attributes.
+   * 
+   * Several of the data structure in this index are in heap, therefore it is
```

Comment:
From my understanding, when you leave out the tag, any text separated from @brief after a newline is converted into @details.